core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/AwsS3ProtocolFactory.java

```diff
@@ -57,4 +73,48 @@ public AwsS3ProtocolFactory build() {
             return new AwsS3ProtocolFactory(this);
         }
     }
+
+    @Override
+    public <T extends AwsResponse> HttpResponseHandler<Response<T>> createCombinedResponseHandler(
+        Supplier<SdkPojo> pojoSupplier, XmlOperationMetadata staxOperationMetadata) {
+
+        return createErrorCouldBeInBodyResponseHandler(pojoSupplier, staxOperationMetadata);
+    }
+
+    private <T extends AwsResponse> HttpResponseHandler<Response<T>> createErrorCouldBeInBodyResponseHandler(
+        Supplier<SdkPojo> pojoSupplier, XmlOperationMetadata staxOperationMetadata) {
+
+        return new AwsXmlPredicatedResponseHandler<>(r -> pojoSupplier.get(),
+                                                     createResponseTransformer(pojoSupplier),
+                                                     createErrorTransformer(),
+                                                     findErrorInS3Response(),
+                                                     staxOperationMetadata.isHasStreamingSuccessResponse());
+    }
+
+    private Function<AwsXmlUnmarshallingContext, AwsXmlUnmarshallingContext> findErrorInS3Response() {
```

Comment:
Can we make it an inner class instead? Because it's difficult to debug issues in a large function and a well-named inner class is more readable and testable, to me.

Reply:
I agree (especially the testable part)!! I'll do that. And write tests.